base/utf16.jl

```diff
@@ -20,92 +45,248 @@ function endof(s::UTF16String)
+            # output surrogate pair
+            buf[out += 1] = UInt16(0xd7c0 + (ch >>> 10))
+            buf[out += 1] = UInt16(0xdc00 + (ch & 0x3ff))
         end
     end
-    push!(buf, 0) # NULL termination
+    @inbounds buf[out + 1] = 0 # NULL termination
     UTF16String(buf)
 end
 
-utf16(x) = convert(UTF16String, x)
-convert(::Type{UTF16String}, s::UTF16String) = s
-convert(::Type{UTF16String}, s::AbstractString) = encode16(s)
-convert(::Type{Array{UInt16,1}}, s::UTF16String) = s.data
-convert(::Type{Array{UInt16}}, s::UTF16String) = s.data
+"
+Converts a UTF8String to a UTF16String
 
-# TODO: optimize this
-convert(::Type{UTF8String}, s::UTF16String) =
-    sprint(length(s.data)-1, io->for c in s; write(io,c::Char); end)
+### Input Arguments:
+*   ::Type{UTF16String}
+*   str::UTF8String
 
-sizeof(s::UTF16String) = sizeof(s.data) - sizeof(UInt16)
-unsafe_convert{T<:Union(Int16,UInt16)}(::Type{Ptr{T}}, s::UTF16String) =
-    convert(Ptr{T}, pointer(s))
+### Returns:
+*   UTF16String
 
-function isvalid(::Type{UTF16String}, data::AbstractArray{UInt16})
-    i = 1
-    n = length(data) # this may include NULL termination; that's okay
-    while i < n # check for unpaired surrogates
-        if utf16_is_lead(data[i]) && utf16_is_trail(data[i+1])
-            i += 2
-        elseif utf16_is_surrogate(data[i])
-            return false
+### Throws:
+*   UnicodeError
+"
+function convert(::Type{UTF16String}, str::UTF8String)
+    dat = str.data
+    # handle zero length string quickly
+    sizeof(dat) == 0 && return empty_utf16
+    # Check that is correct UTF-8 encoding and get number of words needed
+    len, flags, num4byte = check_string(dat)
+    len += num4byte
+    buf = Vector{UInt16}(len+1)
+    @inbounds buf[len+1] = 0
+    # Optimize case where no characters > 0x7f
+    flags == 0 && @inbounds return UTF16String(copy!(buf, dat))
+    out = 0
+    pos = 0
+    @inbounds while out < len
+        ch::UInt32 = dat[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle range 0x80-0x7ff
+        elseif ch < 0xe0
+            buf[out += 1] = ((ch & 0x1f) << 6) | (dat[pos += 1] & 0x3f)
+        # Handle range 0x800-0xffff
+        elseif ch < 0xf0
+            pos += 2
+            buf[out += 1] = get_utf8_3byte(dat, pos, ch)
+        # Handle range 0x10000-0x10ffff
         else
-            i += 1
+            pos += 3
+            ch = get_utf8_4byte(dat, pos, ch)
+            # output surrogate pair
+            buf[out += 1] = UInt16(0xd7c0 + (ch >>> 10))
+            buf[out += 1] = UInt16(0xdc00 + (ch & 0x3ff))
         end
     end
-    return i > n || !utf16_is_surrogate(data[i])
+    UTF16String(buf)
 end
 
+"
+Converts a UTF-16 encoded vector of UInt16 to a UTF8String
+
+### Input Arguments:
+*   ::Type{UTF8String}
+*   dat::Vector{UInt16}
+
+### Returns:
+*   UTF8String
+
+### Throws:
+*   UnicodeError
+"
+function convert(::Type{UTF8String}, dat::Vector{UInt16})
```

Comment:
That is why the setup code is different.  When you have a `Vector{UInt16}` that needs to be converted, there is no explicit `\0` word, so you check for sizeof(dat) to be 0 to see if it is empty,
but when you have a UTF16String, there must be an explicit `\0` which you have to account for.

The integer/char vectors to/from string type conversions are actually some of the most useful of the conversions, because you need them when taking data from a database record, or something passed in via ZMQ, for example.
I think @quinnj 's `String` package so far is the best path I've seen to removing these inconsistencies, and also making some of the conversions less necessary.
Another issue is that some of these conversions using integer/char vectors are already part of Base, so removing them might break packages... (I'm trying to determine what string functionality is actually in use, at least in Base itself or in the registered packages).
For example, would anyone object to a PR removing `RopeString`?